faunadb-common/src/main/java/com/faunadb/common/Connection.java

```diff
@@ -358,90 +352,82 @@ public void syncLastTxnTime(long newTxnTime) {
-
-    byte[] jsonBody = json.writeValueAsBytes(body);
-    request.content().clear().writeBytes(jsonBody);
-
-    request.headers().set(HttpHeaderNames.CONTENT_LENGTH, jsonBody.length);
-    request.headers().set(HttpHeaderNames.CONTENT_TYPE, "application/json; charset=utf-8");
-
-    return request;
+  public static URI appendUri(URI oldUri, String queryKey, String queryValue) throws URISyntaxException {
+    String urlEncodedKey = URLEncoder.encode(queryKey, StandardCharsets.UTF_8);
+    String urlEncodedValue = URLEncoder.encode(queryValue, StandardCharsets.UTF_8);
+    String query = urlEncodedKey + "=" + urlEncodedValue;
+    return new URI(oldUri.getScheme(), oldUri.getAuthority(), oldUri.getPath(),
+            oldUri.getQuery() == null ? query : oldUri.getQuery() + "&" + query, oldUri.getFragment());
   }
 
-  private CompletableFuture<FullHttpResponse> performRequest(final FullHttpRequest request, final Optional<Duration> requestQueryTimeout) {
+  private CompletableFuture<HttpResponse<String>> performRequest(String httpMethod, String path, Optional<JsonNode> body,
+                                                                 Map<String, String> params, final Optional<Duration> requestQueryTimeout) throws Exception {
     final Timer.Context ctx = registry.timer("fauna-request").time();
-    final CompletableFuture<FullHttpResponse> rv = new CompletableFuture<>();
+    final CompletableFuture<HttpResponse<String>> rv = new CompletableFuture<>();
 
-    request.headers().add("Authorization", authHeader);
-    request.headers().set("X-FaunaDB-API-Version", API_VERSION);
+    URI requestUri = URI.create(mkUrl(path));
 
-    if(jvmDriver != null) {
-      request.headers().set(X_FAUNA_DRIVER, jvmDriver.toString());
+    // Encode all query parameters
+    for (Map.Entry<String,String> entry : params.entrySet()) {
+      requestUri = appendUri(requestUri, entry.getKey(), entry.getValue());
     }
 
-    // TODO: replace as follows when upgrade to Java 9 or higher:
-    // requestQueryTimeout.or(queryTimeout)
-    //   .ifPresent(timeout -> request.headers().set(X_QUERY_TIMEOUT, timeout.toMillis()));
-    queryTimeout
-      .ifPresent(timeout -> request.headers().set(X_QUERY_TIMEOUT, timeout.toMillis()));
+
+    HttpRequest.BodyPublisher bodyPublisher = HttpRequest.BodyPublishers.noBody();
+    if (body.isPresent()) {
+      byte[] jsonBody = json.writeValueAsBytes(body.get());
+      bodyPublisher = HttpRequest.BodyPublishers.ofByteArray(jsonBody);
+    };
 
     // If a query timeout has been given for the current request,
     // override the one from the Connection if any
-    requestQueryTimeout
-      .ifPresent(timeout -> request.headers().set(X_QUERY_TIMEOUT, timeout.toMillis()));
-
-    long time = getLastTxnTime();
-    if (time > 0) {
-      request.headers().set("X-Last-Seen-Txn", Long.toString(time));
-    }
-
-    request.retain();
+    Optional<Duration> queryTimeout = requestQueryTimeout.or(() -> connectionQueryTimeout);
```

Comment:
using the connection query timeout as a default seems incorrect. The default request timeout and the connection timeout should be two separate configuration params.

Reply:
I think I'm not entirely following the changes suggested here, so I will just describe the way it works now to make sure we're all on the same page. So, the query timeout is an optional param that can be set both when creating the Connection and when sending individual new requests.  Since it can be set from different places, it takes the following order:

1. Request query timeout (optional, if not set, fallback to the next one)
2. Connection query timeout (optional, if not set, fallback to the next one)
3. Server query timeout

Just to give some context, this is the way it was working before with the previous implementation (the Netty client). When we implemented it at that time, we followed the design described at [DRV-86](https://faunadb.atlassian.net/browse/DRV-86).

Let me know in case above is still correct or not! :)

Reply:
whoops, I misread `connectionQueryTimeout` as `connectionTimeout`... I think I prefer the original name, or something like `defaultQueryTimeout` if we're going to rename it. ("Connection" as a name for the class is a bit of a misnomer already, so better not to propagate that into param names.)

Reply:
I find the `Connection` class not super clear as well so I went with the proposed renaming
https://github.com/fauna/faunadb-jvm/pull/270/commits/917090342ef065e90ce8cc1aea2893a01d1a39ae